src/passes/OptimizeInstructions.cpp

```diff
@@ -3549,6 +3552,29 @@ struct OptimizeInstructions
     return nullptr;
   }
 
+  // Bitwise AND of a boolean result (0 or 1) with any value
+  // whose LSB is guaranteed to be 0 always yields 0.
+  Expression* optimizeAndBooleanWithEvenConstant(Binary* curr) {
+    assert(curr->op == AndInt32);
+
+    using namespace Abstract;
+    using namespace Match;
+    auto leftMaxBits = Bits::getMaxBits(curr->left, this);
+    auto type = curr->left->type;
+    if (curr->op == Abstract::getBinary(type, And)) {
+      if (leftMaxBits == 1) {
+        // boolean & (No overlap with boolean's LSB)  ==> 0
+        if (auto* c = curr->right->dynCast<Const>()) {
+          if ((1 & c->value.getInteger()) == 0) {
+            replaceCurrent(getDroppedChildrenAndAppend(
+              curr, LiteralUtils::makeZero(c->value.type, *getModule())));
+          }
+        }
+      }
+    }
+    return nullptr;
```

Comment:
This could be useful for more than booleans. A more general rule could be `x & c == 0` when `c` has no bits in the range that `x` does, that is, if `x` has bits set in the range `[0, n)` (bit 0 up to bit n-1, but not n), and `c` has bits in `[n, max)`, then the result is 0.